### ISSUE TITLE
Multidimensional condition for conditional invertible neural network

### ISSUE THREAD

**Issue by pauliusinc** [VLL-HD/FrEIA#42]
Hi there!

I am currently trying to build a conditional neural network regarding the MNIST example.
However, I do not want to use the standard procedure to train the MNIST cINN from a 28x28 picture/array to onehot(digit), but I want to train the reverse direction from onehot(digit) to a 28x28 picture/array.

Afterward, I want to be able to reverse sample from a 28x28 picture/array to the onehot(digit).
I have a similar issue to this with larger dimensions and wanted to start simple with the MNIST minimal example.

If I swap the direction I have an input of dimension 1 (one digit) or 10 (digit encoded as onehot) and my condition should be (1, 28, 28). I wanted to ask if it is possible to handle those dimensions for the FrEIA cINN architecture?

I already tried this but I am getting dimension problems similar to this issue:
https://github.com/VLL-HD/FrEIA/issues/9

After taking a deeper look into the error/code I can see there is the following assertion defined:

```
assert all([tuple(dims_c[i][1:]) == tuple(dims_in[0][1:]) for i in range(len(dims_c))]), \
            F"Dimensions of input and one or more conditions don't agree: {dims_c} vs {dims_in}."
```

`FrEIA\modules\coupling_layers.py", line 161`

So, that means if the dimension of the condition(s) is larger than a single 1nd. The input dimensions from index 1+ must be equal. Regarding my example, if I got the condition (1, 28, 28) my input must be (x, 28, 28), right?

What can I do to fit my expectations? I will try to bypass this by scaling my input up to (1, 28, 28) and concatenate all entries with my label or I will try to flat my condition (1, 28, 28) to a 1-dimensional one with a dimension of (784).

Can you recommend me other solutions? It might be possible to split the conditional image into several conditions or not?

Thanks for your help in advance!

Best Regards
Pauliusinc

**Comment by ardizzone** [comment]
Hi!
Sorry for the late response!
We are working on a better tutorial, that will hopefully make things clearer in the future. 
As for your question:
Indeed, with the standard formulation of coupling blocks, a 1D input needs a 1D condition, etc.
I think there are two possibilities:

1.) The easy solution: Just make the condition shape 768 instead of 1, 28, 28, and flatten the MNIST image before inputting it as a condition.

2.) The probably better solution: Add a very simple 'conditioning network', as per the colorization for example.
```python
cond_net = nn.Sequential(nn.Conv2d(1, 32, kernel_size=3),
                                      nn.Relu(),
                                      nn.Conv2d(1, 32, kernel_size=3),
                                      nn.MaxPool2d(2),
                                      nn.Flatten())
# then during training:
c_processed = cond_net(c)
z = inn(x, c_processed)
```
In that way, the cond_net can extract useful features from the condition image.
Then, the condition should be size ((28 - 2 - 2) // 2)**2 = 144 features.
See also our updated conference version of the cINN paper (still to appear in the GCPR proceedings)
https://drive.google.com/drive/folders/1cX2B55B6iDf1zwCASQ-ezmDyCEjNQZWL?usp=sharing
There are some interesting connections to mutual information in there.

If you can't fix the problem with flattening, feel free to re-open the issue!
